numpy/core/getlimits.py

```diff
@@ -485,31 +486,18 @@ def __new__(cls, dtype):
         except TypeError:
             # In case a float instance was given
             dtype = numeric.dtype(type(dtype))
+            # changed type, so we can recursively call finfo
+            return finfo(dtype)
```

Comment:
Would it be so bad to do something like:
```
def __new__(cls, dtype):
    # dtype is None check...
    # try/except `dtype` coercion
    return cls._new(dtype)

@classmethod  # or just a function/static
@lru_cache
def _new(cls, dtype):
    sctype = dtype.type
    if complex:
        # recurse
        dtype = _convert_complex_to_real_float[dtype]
        return cls._new(dtype)
```
Then I don't have to worry about the `lru_cache` requiring `typed=True` (which I am still absolutely certain it would at the very least).  Sure, that doesn't give you bonus speed for some inputs, but it certainly solves the complex caching problem.

Reply:
@seberg Your suggestion is basically version 3 (https://github.com/numpy/numpy/compare/main...eendebakpt:numpy:linalg_performance_v3). Performance results I posted are hardly better than main, although they look a bit suspicious now, so I have to double check that.

But the main question I still have is whether it is ok to loose some of the singleton behaviour of the current `finfo`. Any solution involving with `lru_cache` without the `type=True` has this problem
```
f1 = np.finfo(np.float64)
# ... do some work that fills the lru cache and drops out f1
....
f2 = np.finfo(np.float64)
assert f1 is f2 # will be False! but maybe not a problem?
```
With `type=True` I think we can preserve the singleton behaviour, but it requires some careful analysis.

So my preferred approach would be:

* If we can drop the singleton behaviour, we select a version of the lru_cache implementation that is simple and fast
* If we want to keep the singleton behaviour, I perfer the first implementation https://github.com/numpy/numpy/pull/23088 (which is closest to the current main)

Reply:
We can probably drop the singleton behavior, but I really don't see why it would be hard to keep it.  I don't trust the version here (adding top level lru) without `typed=True` as I said a few times.

The first solution *also* has the same issue as the lru cache without `typed=True`: `float32(1.0)` and `float64(1.0)` are identical based on value and hash and must not be.

Yes, maybe the third version is basically what I said here, in which case it seems to me like it is the best (and currently only viable) solution.